### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by Pinchett-i**

### CODE REVIEW

File: policies/authorize.mjs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import Matrix from './matrix.json' assert { type: "json" };
```
Comment (RoelandMatthijssens) [inline policies/authorize.mjs:1]: at least call it `permission_matrix.json`

File: policies/authorize.mjs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import User from '../models/user.mjs'
+
+async function authorize(controller, action, current_user_id, _records) {
```
Comment (RoelandMatthijssens) [inline policies/authorize.mjs:4]: remove the `_xxx`

File: policies/authorize.mjs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  let conditions = Matrix[controller][action]
```
Comment (RoelandMatthijssens) [inline policies/authorize.mjs:5]: potential nullpointer

**Review by RoelandMatthijssens — CHANGES_REQUESTED** [review]

File: controllers/projects_controller.mjs [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
 class ProjectsController extends ApplicationController {
 
   static async index(request, response) {
-    let projects = await Project.all()
+    let authorized = await authorize('projects', 'index', request.session.current_user_id)
+    if (authorized == false) {
+      request.flash("error", "Unauthorized")
+      response.redirect('/')
+      return
+    }
```
[the remark this answers is not in the document]
Comment (Pinchett-i) [reply controllers/projects_controller.mjs:13]:
```js
static async index(request, response) {
  let authorized = await authorize('projects', 'show', request.session.current_user_id)
  if (!authorized) {
    handle_unauthorized(request, response);
    return
  }
}
      
static async handle_unauthorized(request, response) {
  request.flash("error", "Unauthorized")
  response.redirect('/')
}
```
     
     
     OR 
     
```js
static async index(request, response) {
  this.handle_authorization(request, response, 'projects', 'index')
}

static async handle_authorization(request, response, controller_name, action) {
  let authorized = await authorize(controller_name, action, request.session.current_user_id)
  if (!authorized) {
    request.flash("error", "Unauthorized")
    response.redirect('/')
  }
}
```

**Review by RoelandMatthijssens — APPROVED** [review]

**Pull request merged**